network_simulator/statistics.py

```diff
@@ -4,6 +4,200 @@
+
+        :param Flow flow: flow to be determined.
+        :return :class:`.FlowStats`.
+        """
+        if flow.flow_id not in self.flow_stats:
+            self.flow_stats[flow.flow_id] = FlowStats()
+        return self.flow_stats[flow.flow_id]
+
+    def get_host_stats(self, host):
+        """
+        Check whether a host has already existed in the dictionary; if so,
+        skip the rest of the steps and return the stats object. If not,
+        create a new mapping then return the stats object.
+
+        :param Host host: host to be determined.
+        :return :class:`.HostStats`.
+        """
+        if host.address not in self.host_stats:
+            self.host_stats[host.address] = HostStats()
+        return self.host_stats[host.address]
+
+    def link_buffer_occ_change(self, link, curr_time):
+        """
+        Update the buffer size as a link gains or loses a queue element.
+
+        :param Link link: link that has a change of buffer size.
+        """
+        stats = self.get_link_stats(link)
+        # TODO(sharon): Check with Cody if qsize() would work in this case.
+        buffer_occ_packets = link.buffer.qsize()
+        stats.buffer_occupancy.append((curr_time, buffer_occ_packets))
+
+    def link_packet_loss(self, link, curr_time):
+        """
+        Record a packet loss of a link.
+
+        :param Link link: link of action.
+        :param float curr_time: simulation time of loss.
+        """
+        stats = self.get_link_stats(link)
+        stats.packet_loss_times.append(curr_time)
+
+    def link_packet_transmitted(self, link, packet, curr_time):
+        """
+        Record a packet transmission of a link.
+
+        :param Link link: link of action.
+        :param int packet_size: the size in bits of packet transmitting.
+        :param float curr_time: simulation time of transmission.
+        """
+        stats = self.get_link_stats(link)
+        stats.packet_transmit_times.append((curr_time, packet_size))
+
+    def flow_packet_sent(self, flow, curr_time):
+        """
+        Record a packet sent from a flow.
+
+        :param Flow flow: flow of action.
+        :param float curr_time: simulation time of sending.
+        """
+        stats = self.get_flow_stats(flow)
+        stats.packet_sent_times.append(curr_time)
+
+    def flow_packet_received(self, flow, curr_time):
+        """
+        Record a packet received by a flow.
+
+        :param Flow flow: flow of action.
+        :param float curr_time: simulation time of reception.
+        """
+        stats = self.get_flow_stats(flow)
+        stats.packet_rec_times.append(curr_time)
+        # Retrieve the last element on the sent_times list to find
+        # the corresponding sent time and using the diff we can
+        # determine the rtt time.
+        sent_time = stats.packet_sent_times[-1]
+        stats.packet_rtts.append(curr_time - sent_time)
+
+    def host_packet_sent(self, host, packet_size, curr_time):
```

Comment:
Again, you can pass in the packet itself instead of packet_size and curr_time, and then the packet_sent_times element will depend on the Packet's start_time_sec and the Packet's size_bits. This is probably cleaner.